core/src/com/fourinachamber/fortyfive/map/events/RandomCardSelection.kt

```diff
@@ -0,0 +1,220 @@
+ */
+object RandomCardSelection {
+
+    private var types: Map<String, List<CardChange>> = mapOf()
+    private var biomes: Map<String, Map<String, List<CardChange>>> = mapOf()
+
+    /**
+     * the path to the file from which the data is read from
+     */
+    private const val TYPES_FILE_PATH: String = "maps/events/cardSelectionTypes.onj"
+
+
+    private val cardSelectionSchema: OnjSchema by lazy {
+        OnjSchemaParser.parseFile(Gdx.files.internal("onjschemas/cardSelectionTypes.onjschema").file())
+    }
+
+    fun init() {
+        val file = OnjParser.parseFile(Gdx.files.internal(TYPES_FILE_PATH).file())
+        cardSelectionSchema.assertMatches(file)
+        file as OnjObject
+        val allTypes: MutableMap<String, List<CardChange>> = mutableMapOf()
+        val allTypesOnj = file.get<OnjArray>("types").value.map { it as OnjObject }
+        allTypesOnj.forEach {
+            val name = it.get<String>("name")
+            val changes = it.get<OnjArray>("cardChanges")
+                .value
+                .map { e -> e as OnjObject }
+                .map { e -> CardChange.getFromOnj(e) }
+            allTypes[name] = changes
+        }
+        this.types = allTypes
+
+
+        val allBiomes: MutableMap<String, Map<String, List<CardChange>>> = mutableMapOf()
+        val allBiomesOnj = file.get<OnjArray>("biomes").value.map { it as OnjObject }
+        allBiomesOnj.forEach { biome ->
+            val name = biome.get<String>("name")
+            val tempMap: MutableMap<String, List<CardChange>> = mutableMapOf()
+            biome.value.filter { t -> t.key != "name" }.forEach {
+                val changes = (it.value as OnjArray).value
+                    .map { e -> e as OnjObject }
+                    .map { e -> CardChange.getFromOnj(e) }
+                tempMap[it.key] = changes
+            }
+            allBiomes[name] = tempMap
+        }
+        this.biomes = allBiomes
+    }
+
+    /**
+     * @param cards the cards to take as source (can have doubles)
+     * @param changeNames the names of the types you want to apply to
+     * @param cardsMaxOnce if true, then all chosen cards will be removed from the remaining selection and can therefore be there only once
+     * @param nbrOfCards how many cards you want. However, there can be fewer cards remaining after applying the effects
+     * @throws Exception if a type is not known
+     */
+    fun <T> getRandomCards(
+        cards: List<T>,
+        changeNames: List<String>,
+        cardsMaxOnce: Boolean = false,
+        nbrOfCards: Int,
+        rnd: Random,
+        biome: String,
+        occasion: String,
+    ): List<T> {
+        if (nbrOfCards >= cards.size && cardsMaxOnce) return cards
+        val (tempCards, tempChances) = getCardsWithChances(cards.toMutableList(), changeNames, biome, occasion)
+        return getCardsFromChances(nbrOfCards, tempCards, tempChances, rnd, cardsMaxOnce)
+    }
+
+    fun <T> getCardsFromChances(
```

Comment:
What is the difference between `getCardsFromChances` and `getCardsWithChances`? I don't really understand what any of these functions do